feat: add opentelemetry-instrumentation-threading library

changemyminds opened the pull request:
# Description
At the current stage, Python OpenTelemetry does not support thread instrumentation. It is unable to pass the correct context in the execution thread context, leading to broken trace-IDs when users invoke threads. This results in the inability to properly trace operations.

Fixes #737

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

I have written the `test_threading.py` file for testing, which includes `threading.Thread` and `threading.Timer`, and have also used `futures.ThreadPoolExecutor` for testing.

- [x] tox -e test-instrumentation-threading

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

aabmass commented:
@changemyminds thanks for the PR. There was a lot of discussion on the other PR that was never merged https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1582. Can you explain how this PR is different or if there is something fundamentally wrong with the other PR? It's been a while so we could all use a refresher

changemyminds commented:
> @changemyminds thanks for the PR. There was a lot of discussion on the other PR that was never merged #1582. Can you explain how this PR is different or if there is something fundamentally wrong with the other PR? It's been a while so we could all use a refresher

I think the goal of both PRs is the same: they aim to pass the correct context into the execution thread context, but there are some differences.

- PR #1582 uses a Monkey Patch on the Thread class.
- This PR applies a Monkey Patch via the Thread method and takes the threading timer into consideration.

I don't know why PR #1582 wasn't merged; that's why I have submitted a new PR. I am trying to make improvements with this one. Should I still wait for another PR? I have used this feature in my project, and it works well.

aabmass commented:
@srikanthccv you put a lot of effort into reviewing https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1582. Are you OK with moving forward with this PR since that one has gone stale?

srikanthccv commented:
I am fine with this. The original author may not be able to take it forward so it's for the best to have someone take over that work and continue. I'd make them co-authors since they spent a non-trivial amount of work on it.

changemyminds commented:
@srikanthccv should I also migrate the unit test case from #1582 to this PR?

srikanthccv commented:
If it adds value and is not already covered by another test please do.

changemyminds commented:
@srikanthccv Done. Please take a look thanks.
![image](https://github.com/open-telemetry/opentelemetry-python-contrib/assets/13563347/404843c5-e801-461d-8194-c62670cedfbc)

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/README.rst:
```diff
@@ -0,0 +1,23 @@
+   :target: https://pypi.org/project/opentelemetry-instrumentation-threading/
+
+This library provides instrumentation for the `threading` module to ensure that
+the OpenTelemetry context is propagated across threads.
```
Can you add a note that this instrumentation does not produce any telemetry?

changemyminds replied:
Done

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+[project]
+name = "opentelemetry-instrumentation-threading"
+dynamic = ["version"]
+description = "Threading tracing for OpenTelemetry"
```
nit this description makes it sound like the plugin creates spans. Can you reword to be more general?

changemyminds replied:
Added more description.

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/src/opentelemetry/instrumentation/threading/__init__.py:
```diff
@@ -0,0 +1,102 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
```
Similar to above comment can you just add an explicit note that it does not produce any telemetry?

changemyminds replied:
Done

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/src/opentelemetry/instrumentation/threading/__init__.py:
```diff
@@ -0,0 +1,102 @@
+
+    @staticmethod
+    def __wrap_threading_start(call_wrapped, instance, args, kwargs):
+        span = trace.get_current_span()
+        instance._otel_context = trace.set_span_in_context(span)
```
Can you copy the whole current context instead of recreating a new one? This will be missing other context keys, like baggage or user specified keys.

```py
instance._otel_context = context.get_current()
```

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/tests/test_threading.py:
```diff
@@ -0,0 +1,185 @@
+            expected_trace_id = span_context.trace_id
+            expected_span_id = span_context.span_id
+
+            with futures.ThreadPoolExecutor(max_workers=1) as executor:
+                future = executor.submit(self.fake_func)
+                future.result()
```
I'm not sure if this work correctly after the first call to `submit()`. `ThreadPoolExecutor` lazily spawns threads but the thread will remain running after the first call.

So later calls to `submit()` would have the same context as the first call to `submit()` which started the given thread in the pool, not the context from the latest caller

changemyminds replied:
You are right. Should I also support the ThreadPoolExecutor.submit() method, which can pass the OpenTelemetry context?

changemyminds replied:
I have supported the `ThreadPoolExecutor.submit()` method, please take a look.

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/tests/test_threading.py:
```diff
@@ -0,0 +1,185 @@
+                self.assertEqual(self.global_trace_id, expected_trace_id)
+                self.assertEqual(self.global_span_id, expected_span_id)
+
+    def fake_func(self):
```
Can you refactor this test function and it's state into a separate class or mock to avoid `self.global_*` variables being re-used across test cases? They are not being reset right now and imo it's cleaner to recreate the test data in each test case

changemyminds replied:
Good ideal, thanks! I have created a `MockContext` class to do that.

changemyminds replied:
I used the `self._mock_span_contexts: List[trace.SpanContext] = []` to simplify mock class.

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/src/opentelemetry/instrumentation/threading/__init__.py:
```diff
@@ -0,0 +1,160 @@
+            context.detach(token)
+
+    @staticmethod
+    def __wrap_thread_pool_submit(call_wrapped, instance, args, kwargs):
+        # obtain the original function and wrapped kwargs
+        original_func = args[0]
+        wrapped_kwargs = {
+            ThreadingInstrumentor.__WRAPPER_KWARGS: kwargs,
+            ThreadingInstrumentor.__WRAPPER_CONTEXT: context.get_current(),
+        }
+
+        def wrapped_func(*func_args, **func_kwargs):
+            original_kwargs = func_kwargs.pop(
+                ThreadingInstrumentor.__WRAPPER_KWARGS
+            )
+            otel_context = func_kwargs.pop(
+                ThreadingInstrumentor.__WRAPPER_CONTEXT
+            )
+            token = None
+            try:
+                token = context.attach(otel_context)
+                return original_func(*func_args, **original_kwargs)
+            finally:
+                context.detach(token)
+
+        # replace the original function with the wrapped function
+        new_args = (wrapped_func,) + args[1:]
+        return call_wrapped(*new_args, **wrapped_kwargs)
```
Would it work to just capture the OTel context in the closure of `wrapped_func` function? That way you don't need to modify the kwargs at all

```py
    def __wrap_thread_pool_submit(call_wrapped, instance, args, kwargs):
        # obtain the original function and calling context
        original_func = args[0]
        otel_context = context.get_current()

        def wrapped_func(*func_args, **func_kwargs):
            token = None
            try:
                token = context.attach(otel_context)
                return original_func(*func_args, **func_kwargs)
            finally:
                context.detach(token)

        # replace the original function with the wrapped function
        new_args = (wrapped_func,) + args[1:]
        return call_wrapped(*new_args, **kwargs)
```

changemyminds replied:
It sound great, thanks for your advice. I have updated it.

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/tests/test_threading.py:
```diff
@@ -13,20 +13,27 @@
 
 
+@dataclass
+class MockContext:
+    span_context: trace.SpanContext = None
```
I don't think this class is needed after all. The SpanContext alreadyt contains the trace and span IDs, so you don't need to record them separately.

changemyminds replied:
I used the `self._mock_span_contexts: List[trace.SpanContext] = []` to simplify mock class.

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/tests/test_threading.py:
```diff
@@ -0,0 +1,203 @@
+
+        expected_span_contexts: List[trace.SpanContext] = []
+        futures_list = []
+        for num in range(max_workers):
```
To "prove" the other bug is gone, you would want at least one thread in the pool executing at least 2 tasks. The easiest way to guarantee that would be `max_workers=1` and submitting at least 2 tasks.

changemyminds replied:
Can I create another test case to do that?

changemyminds replied:
Eventually, I decided to create a new test case for thread pool.
- test_trace_context_propagation_in_thread_pool_with_multiple_workers (original)
- test_trace_context_propagation_in_thread_pool_with_single_worker (newer)

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/tests/test_threading.py:
```diff
@@ -0,0 +1,203 @@
+                future = executor.submit(self.fake_func)
+                futures_list.append(future)
+
+        for future in as_completed(futures_list):
+            future.result()
+
+        # check result
+        self.assertEqual(len(self._mock_span_contexts), max_workers)
+        self.assertEqual(
+            len(self._mock_span_contexts), len(expected_span_contexts)
+        )
+        for index, mock_span_context in enumerate(self._mock_span_contexts):
+            self.assert_span_context_equality(
+                mock_span_context, expected_span_contexts[index]
+            )
```
for the purposes of this testing, it would be much simpler to have the future's result be the span context inside of the submitted function (have `fake_func()` return the span context). Then the order is preserved in the results vs expected

changemyminds replied:
For `fake_func()` method has been used for the `threading.Thread`, would you mind to create another `fak_func()` method to do that or modify original method?

```python
# original
    def fake_func(self) -> trace.SpanContext:
        span_context = trace.get_current_span().get_span_context()
        self._mock_span_contexts.append(span_context)
        return span_context

# other, closure method
    def test_trace_context_propagation_in_thread_pool(self):
        def fake_func():
            return  trace.get_current_span().get_span_context()

       # ...ignore
       for num in range(max_workers):
            with self._tracer.start_as_current_span(f"trace_{num}") as span:
                future = executor.submit(fake_func))
```

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/tests/test_threading.py:
```diff
@@ -0,0 +1,203 @@
+        expected_span_context: trace.SpanContext,
+    ):
+        self.assertEqual(result_span_context, expected_span_context)
+        self.assertEqual(
+            result_span_context.trace_id, expected_span_context.trace_id
+        )
+        self.assertEqual(
+            result_span_context.span_id, expected_span_context.span_id
+        )
```
These checks are already covered by L93 and can be removed

changemyminds replied:
Ok, I think I can remove the `assert_span_context_equality` method and only compare the `span_context`, right?

changemyminds replied:
I have modified it, thanks.

aabmass reviewed (approved):
A few small nits but otherwise LGTM. Thanks for the PR!

@open-telemetry/opentelemetry-python-contrib-maintainers do we want ask contributors of new instrumentations to be CODEOWNERS for new instrumentations?

aabmass commented on instrumentation/opentelemetry-instrumentation-threading/tests/test_threading.py:
```diff
@@ -0,0 +1,232 @@
+        result_span_contexts = [future.result() for future in futures_list]
+
+        # check result
+        self.assertEqual(len(result_span_contexts), max_workers)
+        self.assertEqual(
+            len(result_span_contexts), len(expected_span_contexts)
+        )
+        for index, result_span_context in enumerate(result_span_contexts):
+            self.assertEqual(
+                result_span_context, expected_span_contexts[index]
+            )
```
Is something missing if we just compare them directly?

```suggestion
        self.assertEqual(result_span_contexts, expected_span_contexts)
```

changemyminds replied:
You are right. I made a mistake. I think the Python like other languages such Go or Java, if we compared array or object, they will compare the address. I have updated it, thanks.

aabmass commented:
You can just resolve the conversations once they're addressed. The PR looks good to me

changemyminds commented:
Hi @aabmass could you ask other reviewers to review this PR? If there are no issue, can we merge it? I'm concerned that this PR will be overlooked. Thanks.

srikanthccv commented:
Don't be concerned. It will not be ignored as long as there are no open conversations and the requested feedback is addressed.

changemyminds commented:
> Don't be concerned. It will not be ignored as long as there are no open conversations and the requested feedback is addressed.

Thanks for your replying but should I period update the branch in this PR? Because I need to rebbae the branch again and again to ensure the PR is the newer same with the main btanch.

srikanthccv commented:
You don't need to unless there are some code conflicts that should be resolved which shouldn't be the case since this is a brand new instrumentation.

changemyminds commented:
> You don't need to unless there are some code conflicts that should be resolved which shouldn't be the case since this is a brand new instrumentation.

Got it, thanks

changemyminds commented:
Any update news? Should I help to solve the conflicted branch issue?

srikanthccv commented:
Yes, please fix the conflicts.

changemyminds commented:
Hi @aabmass , @ocelotl , and @lzchen , I apologize for the mistake. I accidentally updated the wrong branch when syncing with the `otel-threading` branch, which caused the previous commit about @ocelotl  to disappear. I'm very sorry for the inconvenience.

I have updated the CHANGELOG.md, fixed the conflicts, and changed the tox.ini test library approach to the pip test-requirements.txt approach. Please help me review it again. Thank you.

phillipuniverse commented:
@ocelotl with the merge of this PR, does that mean that https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1582 is no longer necessary and should be closed?

ocelotl commented:
> @ocelotl with the merge of this PR, does that mean that #1582 is no longer necessary and should be closed?

Right, closed #1582, thanks.

nrajb commented:
Hi All,

Do we have plans to publish package for this?
https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-threading

`pip install opentelemetry-instrumentation-threading`
doesn't yet resolve any version, not even 0.46b0.dev

```
└> pip install opentelemetry-instrumentation-threading                                                                                                                                                                       [23/05/24|11:22am]
ERROR: Could not find a version that satisfies the requirement opentelemetry-instrumentation-threading (from versions: none)
ERROR: No matching distribution found for opentelemetry-instrumentation-threading

and

└> pip install -v "opentelemetry-instrumentation-threading==0.46b0.dev"                                                                                                                                                      [23/05/24|11:15am]
Using pip 23.0.1 from /Users/rajbnag/Downloads/meow-connect/.venv/lib/python3.10/site-packages/pip (python 3.10)
ERROR: Could not find a version that satisfies the requirement opentelemetry-instrumentation-threading==0.46b0.dev (from versions: none)
ERROR: No matching distribution found for opentelemetry-instrumentation-threading==0.46b0.dev
```

phillipuniverse commented:
@nrajb this change is not apart of a release yet but will be part of the next one, 0.46b0.
